Support autoprobing node-security-group for openstack cloud provider

FengyunPan opened the pull request:
1. Support autoprobing node-security-group
2. Support multiple Security Groups for cluster's nodes
3. Fix recreating Security Group for cluster's nodes

This is a part of #50726

**Special notes for your reviewer**:
/assign @anguslees 
/assign @dims 

**Release note**:
```release-note
Support autoprobing node-security-group for openstack cloud provider, Support multiple Security Groups for cluster's nodes.
```

dims commented:
/ok-to-test

xsgordon commented:
/sig openstack

FengyunPan commented:
@anguslees PTAL, thank you.

dims commented:
/retest

dims commented:
/test all

dims commented:
/approve

dims commented:
/retest

dims commented:
/retest

dims commented:
/kind feature

FengyunPan commented:
/test pull-kubernetes-e2e-gce-bazel

dims commented:
@FengyunPan needs rebase

dims commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -913,30 +936,79 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *v1.Serv
+			// create it later
+			lbSecGroupID = ""
+		} else {
+			glog.V(1).Infof("Error occurred finding security group: %s: %v", lbSecGroupName, err)
```
Should this be Errorf?

FengyunPan replied:
Oops, update it soon.

FengyunPan commented:
/test pull-kubernetes-e2e-gce-bazel
/test pull-kubernetes-e2e-gce-etcd3

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -999,25 +1060,65 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *v1.Serv
+			res := neutronports.Update(lbaas.network, portID, update_opts)
+			if res.Err != nil {
+				msg := fmt.Sprintf("Error occured updating port %s for loadbalancer service %s/%s: %v", portID, apiService.Namespace, apiService.Name, res.Err)
+				glog.Errorf(msg)
```
log-and-return

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -913,30 +936,79 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *v1.Serv
-
 		if err != nil {
-			// cleanup what was created so far
-			_ = lbaas.EnsureLoadBalancerDeleted(clusterName, apiService)
```
all the way through this function, you've removed this call to cleanup during error unwind.  That's fine, but that makes it likely that this function will be called multiple times with some first portion of the resources already existing from a previous failed attempt.  With this new approach, we need to make sure that _every_ call to create a resource will correctly/safely ignore the case where the resource already exists.

See `openstack_routes.go` for an alternative pattern.  The routes code accumulates "undo" callbacks, and calls them in reverse order if one of the steps encounters an error.  (see references to `onFailure`)

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -999,25 +1060,65 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *v1.Serv
+			}
+			if len(secGroupRules) != 0 {
+				// Do not add rule when find rules for remote group in the Node Security Group
+				continue
```
How do we know the rule you just found is for the correct port/protocol?
I think the listOpts selector has to match the createOpts almost exactly..

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -999,25 +1060,65 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *v1.Serv
+			// Add the rules in the Node Security Group
+			err = createNodeSecurityGroup(lbaas.network, nodeSecurityGroupID, int(port.NodePort), port.Protocol, lbSecGroupID)
+			if err != nil {
+				return fmt.Errorf("Error occured creating security group for loadbalancer service %s/%s", apiService.Namespace, apiService.Name)
```
Always include `err` in the returned error, or someone looking at the logs will have to guess what actually went wrong.

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -1143,6 +1244,91 @@ func (lbaas *LbaasV2) UpdateLoadBalancer(clusterName string, service *v1.Service
+		for removal := range removals {
+			// Delete the rules in the Node Security Group
+			opts := rules.ListOpts{
+				SecGroupID:    removal,
```
You need to also check/verify port and protocol (and direction?), otherwise you will delete too much.

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -1143,6 +1244,91 @@ func (lbaas *LbaasV2) UpdateLoadBalancer(clusterName string, service *v1.Service
+		}
+
+		for _, nodeSecurityGroupID := range lbaas.opts.NodeSecurityGroupIDs {
+			opts := rules.ListOpts{
```
need to include port/protocol/direction

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -615,6 +617,27 @@ func getSubnetIDForLB(compute *gophercloud.ServiceClient, node v1.Node) (string,
+		}
+
+		for _, securityGroup := range srv.SecurityGroups {
+			for _, securityGroupName := range securityGroup {
```
:thinking: So remind me - when will a node have multiple securityGroups, and why is it appropriate to add LB rules into all of them (rather than somehow finding the "right" one)?

Commit:
Support autoprobing node-security-group for openstack cloud provider

1. Support autoprobing node-security-group
2. Support multiple Security Groups for cluster's nodes
3. Fix recreating Security Group for cluster's nodes

This is a part of #50726

FengyunPan reviewed (commented):
@anguslees thank you.

anguslees commented:
/lgtm
